fdbserver/worker.actor.cpp

```diff
@@ -1855,8 +1863,28 @@ ACTOR Future<Void> workerServer(Reference<IClusterConnectionRecord> connRecord,
 		startRole(Role::WORKER, interf.id(), interf.id(), details);
 		errorForwarders.add(traceRole(Role::WORKER, interf.id()));
 
-		wait(waitForAll(recoveries));
-		recoveredDiskFiles.send(Void());
+		// We want to avoid the worker being recruited as storage or TLog before recoverying it is local files,
+		// to make sure:
+		//   (1) the worker can start serving requests once it is recruited as storage or TLog server, and
+		//   (2) a slow recovering worker server wouldn't been recruited as TLog and make recovery slow.
+		// However, the worker server can still serve stateless roles, and if encryption is on, it is crucial to have
+		// some worker available to serve the EncryptKeyProxy role, before opening encrypted storage files.
+		// Here we allow the worker to first register with cluster controller before local file recovery, but indicates
+		// it cannot serve as storage or TLog, then a second normal register request after local file recovery.
```

Comment:
a nit: 
"Here we allow the worker to first register with cluster controller before local file recovery, but indicates it cannot serve as storage or TLog, then a second normal register request after local file recovery." 
                  may be written as 
"The code allows a worker to first register with the ClusterController to be recruited _only_ as a  Stateless process i.e. it can't be recruited as a SS or TLog process; once the local disk recovery is complete (if applicable), the process re-registers with ClusterController as a Stateful process role.